### PR TITLE
Properly implement function extensionality based on cardinality

### PR DESCRIPTION
**Pull request by ajreynol**

### CODE REVIEW

**Comment by ajreynol** [comment]
This has a significant impact:   +113 -16 on the higher-order judgment day set.

File: src/theory/uf/theory_uf.h [hunk 1]
```diff
@@ -185,31 +188,51 @@ class TheoryUF : public Theory {
...
+  Node getExtensionalityDeq(TNode deq);
+
+  /** applyExtensionality
+   *
+   * Given disequality deq f != g, if not already cached, this sends a lemma of
+   * the form:
    *   f = g V (f k) != (g k) for fresh constant k.
-   * on the output channel.
-   * Return value is the number of lemmas sent.
+   * on the output channel. This is a "extensionality lemma".
```
Comment (HanielB) [inline src/theory/uf/theory_uf.h]: a --> an

File: src/theory/uf/theory_uf.h [hunk 2]
```diff
@@ -185,31 +188,51 @@ class TheoryUF : public Theory {
...
+   *
+   * If we pass a null model m to this function, then we add extensionality
+   * lemmas to the output channel and return the total number of lemmas added.
+   * We only add lemmas for functions whose type is finite.
+   *
+   * If we pass a non-null model m to this function, then we add disequalities
+   * that correspond to the conclusion of extensionality lemmas to the model's
+   * equality engine. We only add disequalities for functions whose type is
+   * infinite. We return 0 if the equality engine of m is consistent after this
+   * call, and 1 otherwise.
```
Comment (HanielB) [inline src/theory/uf/theory_uf.h]: Say why only lemmas for finite or infinite depending on the model being non-null or null.

File: src/theory/uf/theory_uf.cpp [hunk 3]
```diff
@@ -670,20 +690,25 @@ void TheoryUF::eqNotifyDisequal(TNode t1, TNode t2, TNode reason) {
...
+  if (it == d_extensionality_deq.end())
+  {
     TypeNode tn = deq[0][0].getType();
+    std::vector<TypeNode> argTypes = tn.getArgTypes();
     std::vector< Node > skolems;
-    for( unsigned i=0; i<tn.getNumChildren()-1; i++ ){
-      Node k = NodeManager::currentNM()->mkSkolem( "k", tn[i], "skolem created for extensionality." );
+    for (unsigned i = 0, nargs = argTypes.size(); i < nargs; i++)
+    {
+      Node k = NodeManager::currentNM()->mkSkolem(
```
Comment (HanielB) [inline src/theory/uf/theory_uf.cpp]: cache nodemanager

**Pull request merged**